frontend/src/pages/analysis/Network.tsx

```diff
@@ -0,0 +1,606 @@
+
+  /** max # of nodes to display, in order of rank */
+  const [maxNodes, setMaxNodes] = useState(
+    Math.min(Math.floor(hardMaxNodes / 10), results.network.nodes.length),
+  );
+
+  /** min/max of probabilities */
+  const probabilityExtent = d3.extent(
+    results.network.nodes.map((node) => node.probability),
+  );
+
+  /** don't display nodes below this probability */
+  const [minProbability, setMinProbability] = useState(probabilityExtent[0]!);
+
+  /** whether to auto-fit camera to contents every frame */
+  const [autoFit, setAutoFit] = useState(false);
+
+  /** selected node */
+  const [selectedNode, setSelectedNode] = useState<Node>();
+
+  /** node label key to display */
+  const [labelKey, setLabelKey] = useState(labelKeyOptions[0]!.id);
+
+  /** nodes to display */
+  const nodes = useMemo(
+    () =>
+      results.network.nodes
+        .filter((node) => node.probability >= minProbability)
+        .slice(0, maxNodes),
+    [results.network.nodes, minProbability, maxNodes],
+  );
+
+  /** quick node lookup */
+  const nodeLookup = useMemo(
+    () => Object.fromEntries(nodes.map((node, index) => [node.entrez, index])),
+    [nodes],
+  );
+
+  /** links to display */
+  const links = useMemo(
+    () =>
+      results.network.links
+        .filter(
+          (link) => link.source in nodeLookup && link.target in nodeLookup,
+        )
+        .map((link) => ({
+          ...link,
+          sourceIndex: nodeLookup[link.source]!,
+          targetIndex: nodeLookup[link.target]!,
+        })),
+    [nodeLookup, results.network.links],
+  );
+
+  /** legend info */
+  const legendInfo: [number, string, string][] = [
+    [4.5, "Nodes", formatNumber(nodes.length)],
+    [5.5, "Links", formatNumber(links.length)],
+  ];
+  if (selectedNode)
+    legendInfo.push(
+      [7, "Selected Node", ""],
+      [8, "Rank", formatNumber(selectedNode.rank)],
+      [9, "Prob.", formatNumber(selectedNode.probability)],
+      [10, "Entrz.", selectedNode.entrez],
+      [11, "Sym.", selectedNode.symbol],
+      [12, "Name", selectedNode.name],
+      [13, "K/N", selectedNode.knownNovel],
+      [14, "Class", selectedNode.classLabel],
+    );
+
+  /** camera zoom handler */
+  const zoom = useMemo(
+    () =>
+      d3
+        .zoom<SVGSVGElement, unknown>()
+        .scaleExtent([minZoom, maxZoom])
+        /** on zoom/pan */
+        .on("zoom", (event) => {
+          /** if due to direct user interaction (wheel, drag, pinch, etc) */
+          if (event.sourceEvent) setAutoFit(false);
+          /** update zoom camera transform */
+          zoomRef.current?.setAttribute(
+            "transform",
+            event.transform.toString(),
+          );
+        })
+        /** prevent zoom over legend */
+        .filter((event) => !legendRef.current?.contains(event.target)),
+    [],
+  );
+
+  /** fit zoom camera to contents of svg */
+  const fitZoom = useCallback(() => {
```

Comment:
This code basically taken from my greenelab adage and connectivity search apps, which have been well-tested.